File: work/10074/main.py
def max_land(x: int, y: int, s: bool, m: list, memo: dict, memoZeros: dict):
  r: int = 0
  if s and x >= 0 and y >= 0:
    for i in range(x, len(m[y])):
      maxArea: int = 0
      if (y, i, s) in memo:
        maxArea = memo[(y, i, s)]
      else:
        minZeros: int = 0
        k = 1
        for j in range(i, len(m[y])):
          zerosC = zeros(j, y, not s, m, memoZeros)
          minZeros = min(minZeros, zerosC) if minZeros != 0 else zerosC
          if minZeros == 0:
            k = 1
          else:
            maxArea = max(maxArea, minZeros * k) 
            k += 1
        memo[(y, i, s)] = maxArea
      r = max(maxArea, r)
    if x == 0:
      r = max(r, max_land(x, y - 1, s, m, memo, memoZeros))
    else:
      r = max(r, max_land(x - 1, y, not s, m, memo, memoZeros))
  elif not s and y >= 0 and x >= 0:
    for i in range(y, len(m)):
      maxArea: int = 0
      if (i, x, s) in memo:
        maxArea = memo[(i, x, s)]
      else:
        minZeros: int = 0
        k = 1
        for j in range(i, len(m)):
          zerosC = zeros(x, j, not s, m, memoZeros)
          minZeros = min(minZeros, zerosC) if minZeros != 0 else zerosC
          if minZeros == 0:
            k = 1
          else:
            maxArea = max(maxArea, minZeros * k)
            k += 1
        memo[(i, x, s)] = maxArea
      r = max(maxArea, r)
    if y == 0:
      r = max(r, max_land(x - 1, y, s, m, memo, memoZeros))
    else:
      r = max(r, max_land(x, y - 1, not s, m, memo, memoZeros))
  return r

def zeros(x: int, y: int, s: bool, m: list, memoZeros: dict):
  if y < 0 or y >= len(m) or x < 0 or x >= len(m[y]):
    return 0
  if m[y][x] == 1:
    return 0
  if (y, x, s) in memoZeros:
    return memoZeros[(y, x, s)]
  if s:
    r = 1 + zeros(x + 1, y, s, m, memoZeros)
  else:
    r = 1 + zeros(x, y + 1, s, m, memoZeros)
  memoZeros[(y, x, s)] = r
  return r
[m, n] = map(int, input().split())
while m != 0 or n != 0:
  l: list = []
  memo = {}
  memoZeros = {}
  for _ in range(m):
    row: list = list(map(int, input().split()))
    l.append(row)
  print(max_land(n - 1, m - 1, True, l, memo, memoZeros))
  [m, n] = map(int, input().split())

File: work/10074/main.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <tuple>
#include <algorithm>

using namespace std;

struct MemoKey
{
  int x, y, minZ;
  bool s;
  bool operator==(const MemoKey &o) const
  {
    return tie(x, y, s) == tie(o.x, o.y, o.s);
  }
};

struct MemoKeyHash
{
  size_t operator()(const MemoKey &key) const
  {
    return hash<int>()(key.x) ^ hash<int>()(key.y) ^ hash<int>()(key.minZ) ^ hash<bool>()(key.s);
  }
};

int zeros(int x, int y, bool s,
          const vector<vector<int>> &m,
          unordered_map<MemoKey, int, MemoKeyHash> &mZ)
{
  if (y >= m.size() || x >= m[y].size())
    return 0;
  if (m[y][x] == 1)
    return 0;

  MemoKey key = {x, y, s};
  if (mZ.find(key) != mZ.end())
  {
    return mZ[key];
  }

  int result;
  if (s)
  {
    result = 1 + zeros(x + 1, y, s, m, mZ);
  }
  else
  {
    result = 1 + zeros(x, y + 1, s, m, mZ);
  }
  mZ[key] = result;
  return result;
}

int maxArea(int x, int y, bool s,
            int k, int minZeros,
            const vector<vector<int>> &m,
            unordered_map<MemoKey, int, MemoKeyHash> &memo)
{
  if (y >= m.size() || x >= m[y].size())
    return 0;
  unordered_map<MemoKey, int, MemoKeyHash> mZ;
  int zerosC = zeros(x, y, !s, m, mZ);
  int minZ = (minZeros != 0)
                 ? min(minZeros, zerosC)
                 : zerosC;
  MemoKey key = {x, y, minZ, s};
  if (memo.find(key) != memo.end())
  {
    return memo[key];
  }
  int maxA = 0;
  if (minZ == 0)
  {
    maxA = maxArea((s ? x + 1 : x),
                   (s ? y : y + 1), s, 1, minZ, m, memo);
  }
  else
  {
    maxA = max(minZ * k, maxArea((s ? x + 1 : x),
                                 (s ? y : y + 1), s, k + 1, minZ, m, memo));
  }
  memo[key] = maxA;
  return maxA;
}

int max_land(int x, int y, bool s,
             const vector<vector<int>> &m)
{
  int r = 0;
  unordered_map<MemoKey, int, MemoKeyHash> memo;

  if (s && x >= 0 && y >= 0)
  {
    for (int i = x; i < m[y].size(); ++i)
    {
      r = max(r, maxArea(i, y, s, 1, 0, m, memo));
    }
    if (x == 0)
    {
      r = max(r, max_land(x, y - 1, s, m));
    }
    else
    {
      r = max(r, max_land(x - 1, y, !s, m));
    }
  }
  else if (!s && y >= 0 && x >= 0)
  {
    for (int i = y; i < m.size(); ++i)
    {
      r = max(r, maxArea(x, i, s, 1, 0, m, memo));
    }
    if (y == 0)
    {
      r = max(r, max_land(x - 1, y, s, m));
    }
    else
    {
      r = max(r, max_land(x, y - 1, !s, m));
    }
  }
  return r;
}
int main()
{
  int m, n;
  cin >> m >> n;
  while (m != 0 || n != 0)
  {
    vector<vector<int>> grid(m, vector<int>(n));

    for (int i = 0; i < m; ++i)
    {
      for (int j = 0; j < n; ++j)
      {
        cin >> grid[i][j];
      }
    }

    cout << max_land(n - 1, m - 1, true, grid) << '\n';
    cin >> m >> n;
  }
  return 0;
}